docker/docker_image_src.go

```diff
@@ -1,69 +1,92 @@
 }
 
-func parseDockerImage(img, certPath string, tlsVerify bool) (types.Image, error) {
-	s, err := newDockerImageSource(img, certPath, tlsVerify)
+func (e errFetchManifest) Error() string {
+	return fmt.Sprintf("error fetching manifest: status code: %d, body: %s", e.statusCode, string(e.body))
+}
+
+type dockerImageSource struct {
+	ref reference.Named
+	tag string
+	c   *dockerClient
+}
+
+// newDockerImageSource is the same as NewDockerImageSource, only it returns the more specific *dockerImageSource type.
+func newDockerImageSource(img, certPath string, tlsVerify bool) (*dockerImageSource, error) {
+	ref, tag, err := parseDockerImageName(img)
 	if err != nil {
 		return nil, err
 	}
-	return &dockerImage{src: s}, nil
-}
-
-func (i *dockerImage) RawManifest(version string) ([]byte, error) {
-	// TODO(runcom): unused version param for now, default to docker v2-1
-	if err := i.retrieveRawManifest(); err != nil {
+	c, err := newDockerClient(ref.Hostname(), certPath, tlsVerify)
+	if err != nil {
 		return nil, err
 	}
-	return i.rawManifest, nil
+	return &dockerImageSource{
+		ref: ref,
+		tag: tag,
+		c:   c,
+	}, nil
 }
 
-func (i *dockerImage) Manifest() (types.ImageManifest, error) {
-	// TODO(runcom): unused version param for now, default to docker v2-1
-	m, err := i.getSchema1Manifest()
+// NewDockerImageSource creates a new ImageSource for the specified image and connection specification.
+func NewDockerImageSource(img, certPath string, tlsVerify bool) (types.ImageSource, error) {
+	return newDockerImageSource(img, certPath, tlsVerify)
+}
+
+func (s *dockerImageSource) GetManifest() (manifest types.ImageManifest, unverifiedCanonicalDigest string, err error) {
+	url := fmt.Sprintf(manifestURL, s.ref.RemoteName(), s.tag)
+	// TODO(runcom) set manifest version header! schema1 for now - then schema2 etc etc and v1
+	// TODO(runcom) NO, switch on the resulter manifest like Docker is doing
+	res, err := s.c.makeRequest("GET", url, nil, nil)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
-	ms1, ok := m.(*manifestSchema1)
-	if !ok {
-		return nil, fmt.Errorf("error retrivieng manifest schema1")
-	}
-	tags, err := i.getTags()
+	defer res.Body.Close()
+	manblob, err := ioutil.ReadAll(res.Body)
 	if err != nil {
-		return nil, err
+		return nil, "", err
+	}
+	if res.StatusCode != http.StatusOK {
+		return nil, "", errFetchManifest{res.StatusCode, manblob}
 	}
-	imgManifest, err := makeImageManifest(i.src.ref.FullName(), ms1, i.digest, tags)
+	// TODO(remove, already set in manifest): Miloslav?
+	unverifiedCanonicalDigest = res.Header.Get("Docker-Content-Digest")
+	tags, err := s.getTags()
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
-	return imgManifest, nil
+	manifest, err = makeImageManifest(s.ref.FullName(), manblob, unverifiedCanonicalDigest, tags)
+	return
 }
 
-func (i *dockerImage) getTags() ([]string, error) {
+func (s *dockerImageSource) getTags() ([]string, error) {
 	// FIXME? Breaking the abstraction.
```

Comment:
"breaking the abstraction” refers to `dockerImage` directly using members of `i.src` = `dockerImageSource`. That would not be applicable any longer if the method were moved to `dockerImageSource`.